Stargate upgrade proposal (#5)
cosmos/governance: proposals/2020-11-stargate-upgrade/README.md

# Cosmos Stargate Hub Upgrade Proposal 2

![](https://siasky.net/AADU5rg0GvapbrICLUk7SxZeMDrKo5Q8oLrpOwl71wXhnQ)

## by Iqlusion

Jan 11, 2021

## Key Results

------------

> If passed, this governance proposal would commit the Cosmos Hub to halting the `cosmoshub-3` at 06:00 UTC on Jan 28th exporting the state and starting `cosmoshub-4` based on gaia 3.0.

Reviewer: ## Localized times

West Coast USA: 10 PM on Jan 27
East Coast USA: 1 AM on Jan 28
Central Europe: 7 AM on Jan 28
Seoul: 5 PM on Jan 28

Author: Localized times below.

------------

> If passed, this governance proposal would commit the Cosmos Hub to upgrading to halting the `cosmoshub-3` at 06:00 UTC on Nov 19 exporting the state and starting `cosmoshub-4` based on gaia 3.0.

Reviewer 2: ```suggestion
If passed, this governance proposal would commit the Cosmos Hub to upgrading by halting the `cosmoshub-3` at 06:00 UTC on Nov 19, exporting the state, and starting `cosmoshub-4` based on gaia 3.0.
```

## Context

------------

In proposal 27, Iqlusion proposed a comprehensive process to translate the unprecented surface area of this upgrade. We called this process Stargate.

> Iqlusion is pleased to report that the Stargate Process has been successfully executed. We believe the chevrons are unlocked and that Hub can safely step throught the Stargate.

Reviewer 2: ```suggestion
Iqlusion is pleased to report that the Stargate Process has been successfully executed. We believe the chevrons are unlocked and that Hub can safely step through the Stargate.
```

## The Stargate Plan

------------

> The purpose of the Stargate effort was to ensure that Cosmos can fulfill the vision of an Internet of Blockchains in 2020 while mitigating the risks outlined above. It was also an attempt to create a process for complex future on-chain upgrades on the Cosmos Hub. That included the following steps:

Reviewer: 2021

> The purpose of the Stargate effort was to ensure that Cosmos can fulfill the vision of an Internet of Blockchains in 2020 while mitigating the risks outlined above. It was also an attempt to create a process for complex future on-chain upgrades on the Cosmos Hub. That included the following steps:

Reviewer: ```suggestion
The purpose of the Stargate effort was to ensure that Cosmos can fulfill the vision of an Internet of Blockchains in 2021 while mitigating the risks outlined above. It was also an attempt to create a process for complex future on-chain upgrades on the Cosmos Hub. That included the following steps:
```

------------

> The purpose of the Stargate effort was to ensure that Cosmos can fulfill the vision of an Internet of Blockchains in 2020 while mitigating the risks outlined above. It was also an attempt to create a process for complex future on-chain upgrades on the Cosmos Hub. That included the following steps:

Reviewer 2: Reference to "risks outlined above" - presumably should be risks outlined in the original and add some link to it?

Reviewer: We could also just remove this wording.

Reviewer 2: I'd say its important to keep reference to risks as explicit as possible

1. Run the Stargate Testnet and use this as an integration testing target with widest possible ecosystem participation to help confirm the readiness of the release software.

2. Engage and support critical partners during the integration process so that partners are able to actively participate in the testing process and provide insights into their success with the upgrade.

3. Dedicate significant resources in terms of full time human resources and documentation efforts to ensure that everyone integrated into the Cosmos Hub can self-certify as Stargate-ready.

4. Report back to the Hub on the success of the integration process.

   - We expect that the primary responsibility of the Cosmos Hub is to assess whether we have mitigated the above risks sufficiently.

   - Provide a written report of the entire Stargate effort to enable Hub governance to make an informed decision.

5. At conclusion, we will indicate why we have confidence that an upgrade won't be disruptive to the Hub’s ecosystem.

## Stargate Plan Results

------------

### 1. Ecoystem Participation

> The results of ecosystem participation and engagement are available in our [Stargate Ecosystem Readiness Report](https://github.com/cosmosdevs/stargate/blob/master/ecosystem_readiness.md). We list on this page ecosystem partners such as Validators that provided their validator public keys as well as infrastructure partners that confirmed their integration testing the Stargate testnet tags. Validators, exchanges, and wallets have made up the most responsive of their results completing integration testing of the stargate releases.

Reviewer 2: Many of these are still `Integration underway` - I guess that matrix could use an update

Author: This is expected to be a living document. We expect the hard deadline will motivate a lot of teams to finish the work.

### 2. Critical Partner Support

We delivered critical partner support to leading ATOM exchanges. Out of the exchange support efforts, we documented the [IBC Readiness Matrix](https://github.com/cosmosdevs/stargate/blob/master/ibc_readiness_matrix.md) that outlines the levels of readiness that exchanges may select as they upgrade to Cosmos Stargate.

We also completed and delivered a completed legacy Amino Audit, [AminoRest & You](https://github.com/cosmosdevs/stargate/blob/master/audit.md). The core finding of this audit was that changes to the underlying structs result in an interface that is close to the prior version allowing legacy queries to return valid data.

### 3. Documentation and Self-Certification

Documentation was a success for the Cosmos Stargate effort. The upgrades with the most breaking changes such as legacy Amino have a complete audit with documentation on exceptions for blockchain API interfaces.

Most exchanges and wallets have taken self-certification on directly. Our team continues to provide real-time support on multiple partner slack channels and on the Cosmos [#stargate Discord](https://discord.gg/W8trcGV) channel.

> ### Cosmos Stargate Integration Success

Reviewer 2: Do you want to more directly address results for (4) and (5) from the plan?

Reviewer 2: Seems points (4) and (5) from the plan still have no counterpart under results. 

More specifically point (4) of the Plan says `   - Provide a written report of the entire Stargate effort to enable Hub governance to make an informed decision.` - what constitutes that report? 

And (5) says `At conclusion, we will indicate why we have confidence that an upgrade won't be disruptive to the Hub’s ecosystem.`. I can imagine someone implicitly concluding that if they have followed along but it might be good to also summarize this

Cosmos Stargate integration success with exchanges and wallet providers reflects the quality of the code developed by the various Cosmos teams.

## Prop 29

> This upgrade also implements the fund recovery procedure defined in proposal 29.

Reviewer 2: Would be good to link to the prop29 implementation work here so people know what to review re how the genesis file is being generated

## Stargate Upgrade Steps

------------
The upgrade steps for the simulated upgrade of the current Cosmoshub mainnet to Cosmos Stargate includes the following steps:

This section is with the current `gaia 2.0.*` implementation.

>   1. Validators should set their `gaia` with a halt time of `1605765600`ie. 06:00 UTC on Nov 19th in Unix time.

Reviewer 2: Why not define this in terms of height?

Author: Because the height time has a lot of drift in it and makes it very difficult to plan their day more than 12 hours out.

Reviewer 2: Do we forsee any risk in time based? We never did this before right, last time it was height? 

I guess the edge case is if the halt time is towards the end of voting for height H+1. then its possible +1/3 stop at H, after precommiting for H+1 but before committing it, and +1/3 stop at H+1 after committing it. Rare edge case but seemingly possible. 

Correct recovery behaviour here would prob be to resume from H+1 given it did get committed.

  2. Validators should then export the current cosmos state with `gaiad export > cosmoshub-3-export.json`

>   3. Validators should determine the height the last block.

Reviewer 2: ```suggestion
  3. Validators should determine the height of the last block.
```

  1. Validators should then migrate the exported genesis state. `gaiad migrate cosmoshub-3-export.json --chain-id=cosmoshub-4 --initial-height [last_cosmoshub-3_block+1] > genesis.json`
  2. Validators should delete their `~/.gaiad` directory and create a new one with `gaiad init [name]` and then edit their configuration files.
  3. Validators should then start `cosmoshub-4` with `gaiad start`. Automated audits of the genesis state can take 30-120 min using the crisis module. This can be disabled by `gaiad start --x-crisis-skip-assert-invariants`.

## Time of the Upgrade

------------
We propose scheduling the Cosmoshub-3 to Cosmoshub-4 simulated upgrade for Thursday November 19th, 2020 at [0600 UTC]

## The git commit of Gaia that we are upgrading to

------------
TBD

## What we're doing while voting is happening

------------

Testing and integration work will be ongoing with our simulated upgrade testnet running, IBC channels and opened and we will be supporting validator partners.

Large scale simulation testing will be on going.

We have also worked with the teams to double the bug bounty on the stargate upgrade.

> ## What is the upgrade abort process

Reviewer 2: The last upgrade procedure mentioned other abort cases like `need for ad-hoc genesis file changes` or `The failure of cosmoshub-3 to produce two (2) blocks by 180 minutes after the timestamp of Block Height 2902000`.

See language in https://www.mintscan.io/cosmos/proposals/19. Might be worth including something like that here.

Also we can define the COSMOS_HUB_3_FINAL_BLOCK as a variable at the top and just reference that, same for any other vars that might change

------------

> There are multiple circumstances while the proposal should be abandoned even if it passes.34

Reviewer 2: ```suggestion
There are multiple circumstances where the proposal should be abandoned even if it passes:
```

1. A critical vulnerability may be found in the software. If the developement teams change their reccomended version of gaia, the validator set should implicitly abandon this upgrade procedure. A future proposal will be made to the Hub to upgrade to the new target commit.

> 2. The migration process fails could fail to produce a valid cosmoshub-4 genesis file. In this case, the validator set should restart cosmoshub-3 and a future governance proposal will be done to initiate another upgrade.

Reviewer 2: ```suggestion
2. The migration process fails to produce a valid cosmoshub-4 genesis file. In this case, the validator set should restart cosmoshub-3 and a future governance proposal will be done to initiate another upgrade.
```

Reviewer 3: ^ looks like this typo is still here

# Cosmos Stargate Hub Upgrade Proposal 2: Chevrons Locked. Time to Upgrade

![Stargate Upgrade banner graphic](https://siasky.net/AADU5rg0GvapbrICLUk7SxZeMDrKo5Q8oLrpOwl71wXhnQ)

## by Iqlusion

> Jan 11, 2021

Reviewer 3: kind of a nit, but do you want to update the file names for both these files to be `2021-01` rather than `2020-11`?

## Key Results

------------

> If passed, this governance proposal would commit the Cosmos Hub to halting the `cosmoshub-3` at 06:00 UTC on Jan 28th exporting the state and starting `cosmoshub-4` based on gaia 3.0.

Reviewer 4: I want to push for whatever date that is put here is at least 36 hours after expected proposal accept / reject time, using average block time. 

48 hours because variance in expected proposal accept/reject time over 2 weeks could easily alter +- 12 hours, and I think a minimum notice of 24 hours to see the final decision before needing to act upon it is reasonable.

Reviewer 3: How long does voting usually take?

Reviewer 5: @tessr voting period is 2 weeks, same as deposit period

Author: Yes strong agree with min 36 hours window.

## Context

------------

> In proposal 27, Iqlusion proposed a comprehensive process to translate the unprecented surface area of this upgrade. We called this process Stargate.

Reviewer 3: ```suggestion
In [Proposal 27](https://ipfs.io/ipfs/Qmbo3fF54tX3JdoHZNVLcSBrdkXLie56Vh2u29wLfs4PnW), Iqlusion proposed a comprehensive process to translate the unprecented surface area of this upgrade. We called this process Stargate.
```

> Iqlusion is pleased to report that the Stargate Process has been successfully executed. We believe the chevrons are unlocked and that Hub can safely step through the Stargate.

Reviewer 3: What is a chevron?

Author: Stargate SG1 lore.

Reviewer 3: I'd advocate for using less arcane language here, but it's your call

After the success of the community approval of the first Stargate Upgrade proposal, we now follow-up with the second Stargate proposal for the Cosmos Hub that will outline the results of the first proposal and the process for upgrading the Hub.

## The Stargate Plan

------------

> The purpose of the Stargate effort was to ensure that Cosmos can fulfill the vision of an Internet of Blockchains in 2021 while mitigating the risks outlined above. It was also an attempt to create a process for complex future on-chain upgrades on the Cosmos Hub. That included the following steps:

Reviewer 3: > the risks outlined above

Personally, I know what these risks are, but they aren't outlined above in this proposal

   - Provide a written report of the entire Stargate effort to enable Hub governance to make an informed decision.

5. At conclusion, we will indicate why we have confidence that an upgrade won't be disruptive to the Hub’s ecosystem.

## Stargate Plan Results

------------

### 1. Ecoystem Participation

The results of ecosystem participation and engagement are available in our [Stargate Ecosystem Readiness Report](https://github.com/cosmosdevs/stargate/blob/master/ecosystem_readiness.md). We list on this page ecosystem partners such as Validators that provided their validator public keys as well as infrastructure partners that confirmed their integration testing the Stargate testnet tags. Validators, exchanges, and wallets have made up the most responsive of their results completing integration testing of the stargate releases. We anticpate that all actors in the ecosystem are aware and ready to respond to the new Cosmos Hub features. Integration partners have helped find regression in the legacy amino compatibility layer and those have been resolved.

> The Cosmos SDK, Tendermint, AiB, and IBC teams worked tirelessly to deliver the most robust software feature set that would make up the Cosmos Stargate release. Their review of their efforts are documented in their post ["How Seven Teams Collaborated To Deliver The Biggest Software Upgrade In The Cosmos Universe"](https://blog.cosmos.network/how-seven-teams-collaborated-to-deliver-the-biggest-software-upgrade-in-the-cosmos-universe-2288f4f9afe8).

Reviewer 3: ```suggestion
The Cosmos SDK, Tendermint Core, Tendermint Inc, and IBC teams worked tirelessly to deliver the most robust software feature set that would make up the Cosmos Stargate release. Their review of their efforts are documented in their post ["How Seven Teams Collaborated To Deliver The Biggest Software Upgrade In The Cosmos Universe"](https://blog.cosmos.network/how-seven-teams-collaborated-to-deliver-the-biggest-software-upgrade-in-the-cosmos-universe-2288f4f9afe8).
```

maybe?

> Integration testing continued through the release of the Cosmos SDK v0.40.0-rc2 that makes up the Cosmos-hub Stargate testnet tag for a simulated upgrade of the CosmosHub-3 to CosmosHub-4. Also, the AiB team continues to deliver simulation testing of Cosmos Stargate to ensure that any possible chain bug issues can be detected and documented.

Reviewer 3: Again, I think AIB is trying to go by Tendermint Inc, although maybe @alessio should confirm

### Cosmos Stargate Integration Success

Cosmos Stargate integration success with exchanges and wallet providers reflects the quality of the code developed by the various Cosmos teams.

## Prop 29

This upgrade also implements the fund recovery procedure defined in proposal 29. The code is [here](https://github.com/cosmos/gaia/blob/main/app/prop29.go). As requested during the discussion of the governance proposal, the migration and fund recovery procedure verifies all signatures. This code was tested during the cosmoshub-3 simulated upgrade testnet.ƒ

## Stargate Upgrade Steps

------------
The upgrade steps for the simulated upgrade of the current Cosmoshub mainnet to Cosmos Stargate includes the following steps:

This section is with the current `gaia 2.0.*` implementation.

  1. Validators should set their `gaia` with a halt time of `1611813600`ie. 06:00 UTC on Jan 28th in Unix time.

  2. Validators should then export the current cosmos state with `gaiad export > cosmoshub-3-export.json`

  3. Validators should determine the height of the last block as `last-cosmoshub-3_block`.

  4. Validators should back up their `.gaiad` directory.

> This section is with the upgrade `gaia 3.0.*` implemenataion.

Reviewer 3: Should we be more explicit about doing the installation at this point, or does every validator know what this means?

Reviewer 5: yeah it's outlined on the release notes

>   1. Validators should then migrate the exported genesis state. `gaiad migrate cosmoshub-3-export.json --chain-id=cosmoshub-4 --initial-height [last_cosmoshub-3_block+1] > genesis.json`

Reviewer 5: why is this not a zero height upgrade? the last one upgrades reset

Author: We have committed to a number of partners mostly exchanges that we would not reset height during this upgrade. 

Resetting height introduces substantial complexities in their internal systems and we have communicated that this would be unnecessary.

Reviewer 4: I also think zero height upgrades are a much worse UX for users, for both a conceptual front of how they think about the chain, and also for querying historical data. (They just want to query what happened at block x, they shouldn't have to know which hard fork it was in. Noone keeps track of which btc hard fork number we're on rn.)

Reviewer 5: yeah I agree it makes the most sense. I'd just point that out on the gov proposal on a small note at the bottom as the previous upgrades started from 0.

  2. Validators should delete their `~/.gaiad` directory and create a new one with `gaiad init [name]` and then edit their configuration files.
  3. Validators should then start `cosmoshub-4` with `gaiad start`. Automated audits of the genesis state can take 30-120 min using the crisis module. This can be disabled by `gaiad start --x-crisis-skip-assert-invariants`.

Validators should expect that at least 16GB of RAM needs to be provisioned to process block 1 on cosmoshub-4.

## Time of the Upgrade

------------

> We propose scheduling the Cosmoshub-3 to Cosmoshub-4 simulated upgrade for Thursday Jan 28th, 2021 at 0600 UTC

Reviewer 3: Should the Gaia/Cosmos SDK/Tendermint Core engineering teams be on standby at this time?

Reviewer 5: probably not all of them but I think a few ppl would need to monitor the upgrade

- West Coast USA: 10 PM on Jan 27
- East Coast USA: 1 AM on Jan 28
- Central Europe: 7 AM on Jan 28
- Seoul: 5 PM on Jan 28

## The git commit of Gaia that we are upgrading to

------------
`d974b27a8caf8cad3b06fbe4678871e4b0b69a51`

## What we're doing while voting is happening

------------

We expect that integration partners will be motivated to finish their preperations for the upgrade during the period when the governance proposal is running.

A testnet with the final version of Gaia 3.0 will be running to assist with integrations.

[Stargate Ecosystem Readiness Report](https://github.com/cosmosdevs/stargate/blob/master/ecosystem_readiness.md) is living document. We will be updating this document as we get reports of competed end to end testing. We expect a hard deadline to be movtivating.

## What is the upgrade abort process

------------

There are multiple circumstances where the proposal should be abandoned even if it passes:

> 1. A critical vulnerability may be found in the software. If the developement teams change their reccomended version of gaia, the validator set should implicitly abandon this upgrade procedure. A future proposal will be made to the Hub to upgrade to the new target commit.

Reviewer 5: ```suggestion
1. A critical vulnerability may be found in the software. If the development teams change their recommended version of gaia, the validator set should implicitly abandon this upgrade procedure. A future proposal will be made to the Hub to upgrade to the new target commit.
```
